apps/web/src/components/dashboard/peaks/energy-peak-device-assignment-form.tsx

```diff
@@ -12,39 +12,76 @@ import {
     FormItem,
     FormLabel,
     FormMessage,
-    Select,
-    SelectContent,
-    SelectItem,
-    SelectTrigger,
-    SelectValue,
+    MultiSelect,
+    type Option,
 } from "@energyleaf/ui";
 import { zodResolver } from "@hookform/resolvers/zod";
 import { track } from "@vercel/analytics";
+import { useEffect, useState } from "react";
 import { useForm } from "react-hook-form";
 import { toast } from "sonner";
 import type { z } from "zod";
 
 interface Props {
-    devices: DeviceSelectType[];
-    initialValues: z.infer<typeof peakSchema>;
-    sensorId: string;
-    timestamp: string;
+    userId: string;
+    sensorDataId: string;
     onInteract: () => void;
 }
 
-export function EnergyPeakDeviceAssignmentForm({ devices, initialValues, sensorId, timestamp, onInteract }: Props) {
+const deviceSchemaToOption = (devices: z.infer<typeof deviceSchema>[]): Option[] =>
+    devices.map((device) => ({
+        value: device.id.toString(),
+        label: device.name,
+    }));
+
+const optionToDeviceSchema = (option: Option): z.infer<typeof deviceSchema> => ({
+    id: Number(option.value),
+    name: option.label,
+});
+
+export function EnergyPeakDeviceAssignmentForm({ userId, sensorDataId, onInteract }: Props) {
     const form = useForm<z.infer<typeof peakSchema>>({
         resolver: zodResolver(peakSchema),
         defaultValues: {
-            ...initialValues,
+            device: [],
         },
     });
+    const [loadPreselection, setLoadPreselection] = useState(false);
+
+    useEffect(() => {
+        let ignore = false;
+
+        async function getSelectedDevices() {
+            try {
+                setLoadPreselection(true);
+                const devices = await getDevicesByPeak(sensorDataId);
+
+                if (!ignore) {
+                    form.setValue(
+                        "device",
+                        devices.map((device) => ({
+                            id: device.id,
+                            name: device.name,
+                        })),
+                    );
+                }
+            } finally {
+                setLoadPreselection(false);
+            }
+        }
+
+        getSelectedDevices();
+
+        return () => {
+            ignore = true;
+        };
+    }, [form.setValue, sensorDataId]);
```

Comment:
Können wir die Fetches auf dem Client durch Tanstack Query realisieren?
Bietet uns definitiv viele Vorteile (Caching, Stale Validation, Revalidating etc.) auch an anderen Stellen. Für Data Fetching in Client Components, können wir als queryFn auch eine Server Action nehmen.

https://tanstack.com/query/latest/docs/framework/react/quick-start

Reply:
Das habe ich mal für den Fall hier umgesetzt. Gerne Feedback geben

An dieser Stelle ist das meiner Meinung nach zwar ein bisschen overkill, weil man ja eigentlich nichts revalidieren muss, da sich der Dialog ja beim Speichern schließt. Aber ich sehe die Vorteile und für andere Stellen im Code ist das sicherlich sinnvoll.

Reply:
Ist auch eher für die Zukunft, dann haben wir da das Setup.